src/node_file.cc

```diff
@@ -1949,6 +1949,31 @@ static void ReadDir(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+static inline bool CheckOpenPermissions(Environment* env,
```

Comment:
```suggestion
static inline bool CheckOpenPermissions(Environment* env,
```

Any reason to return `bool`? When permission is denied it will throw, so `void` seems appropriate

Reply:
> When permission is denied it will throw, so `void` seems appropriate

It will schedule a pending exception within the JavaScript isolate, but that is unrelated to the C++ control flow. We compile Node.js with `-fno-exceptions`, so any C++ exception would terminate the process.

Reply:
But, technically this function will always return `true`. My point is, why do we need to _early return_ on fs.open and not in other places calling `THROW_IF_INSUFFICIENT_PERMISSIONS`?

Reply:
I'm wondering if we should do it in other places calling `THROW_IF_INSUFFICIENT_PERMISSIONS`

Reply:
I'm not sure what you mean. The `THROW_IF_INSUFFICIENT_PERMISSIONS` macro _always_ returns from the surrounding C++ function when it schedules a JavaScript exception...

https://github.com/nodejs/node/blob/0b328b2e8f5e22a706bde9b879e3c733ce88c865/src/permission/permission.h#L26-L28

While often used with `void` return types, this particular case requires a non-`void` return type.

The implementation of `THROW_IF_INSUFFICIENT_PERMISSIONS` (which _you_ added) specifically accounts for such situations through `__VA_ARGS__`.

Reply:
Sorry, I get it now. Thank you for the explanation.

Reply:
Aren't we supposed to use the `Maybe` type when a function may schedule a JS exception?

Reply:
I do feel differently :slightly_smiling_face: I understand the “this is simple and we don’t _need_ it here logic”, but I do think it’s worth being consistent about this, so that a) code is easier to reason about in general and b) new code is more likely to also follow a consistent pattern. Not saying you have to do this here, but, yeah, I’d prefer it for sure.